src/lib/io/sol-uart-impl-linux.c

```diff
@@ -32,49 +33,126 @@
 #include "sol-uart.h"
 #include "sol-util-internal.h"
 #include "sol-util-file.h"
+#include "sol-vector.h"
 
 SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "uart");
 
+#define DEFAULT_BUFFER_SIZE (4096)
+
 #define FD_ERROR_FLAGS (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP | SOL_FD_FLAGS_NVAL)
 
 struct sol_uart {
+    struct sol_fd *fd_handler;
+    const void *user_data;
+    ssize_t (*rx_cb)(void *data, struct sol_uart *uart, const struct sol_buffer *buf);
+    void (*tx_cb)(void *data, struct sol_uart *uart, struct sol_blob *blob, int status);
+    struct sol_timeout *read_timeout;
+    struct sol_ptr_vector pending_blobs;
+    struct sol_buffer rx;
+    size_t tx_size;
+    size_t pending_tx;
+    size_t written; //How many bytes We've written so far for the first element of pending_blobs
     int fd;
-    struct {
-        void *rx_fd_handler;
-        void (*rx_cb)(void *data, struct sol_uart *uart, uint8_t byte_read);
-        const void *rx_user_data;
-
-        void *tx_fd_handler;
-        void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status);
-        const void *tx_user_data;
-        const uint8_t *tx_buffer;
-        unsigned int tx_length, tx_index;
-    } async;
 };
 
 static bool
-uart_rx_callback(void *data, int fd, uint32_t active_flags)
+read_timeout(void *data)
 {
     struct sol_uart *uart = data;
+    ssize_t r;
+    int err;
+
+    r = uart->rx_cb((void *)uart->user_data, uart, &uart->rx);
+    SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+
+    err = sol_buffer_remove_data(&uart->rx, 0, r);
```

Comment:
as reviewed in the sample code, this will blow if user delete the uart from read callback, like in "I want to read N bytes and then I can close the uart".